File: ops/guard-sql.mjs
/**
 * L4: khong cau SQL nao duoc cham bang memory_fact ngoai src/memory/repository/.
 *
 * dependency-cruiser so khop DUONG DAN MODULE, khong doc duoc chuoi SQL — luat
 * 'memory-fact-chi-o-repository' chi chan viec IMPORT fact.repo, khong chan ai do
 * viet 'SELECT ... FROM memory_fact' thang trong knowledge/. Day la lop con thieu.
 *
 * Chay: npm run guard:sql   (bat buoc trong CI)
 */
import { readdirSync, readFileSync, statSync } from 'node:fs';
import { join, posix, sep } from 'node:path';

const ROOT = 'src';
const ALLOWED_PREFIX = 'src/memory/repository/';
const NEEDLE = 'memory_fact';

/** @param {string} dir @returns {string[]} */
function walk(dir) {
  return readdirSync(dir).flatMap((entry) => {
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) return walk(full);
    return full.endsWith('.ts') ? [full] : [];
  });
}

/**
 * Xoa comment nhung GIU nguyen so dong, de bao loi con chi dung cho.
 * Nhac ten bang trong comment la hop le; chay SQL len no thi khong.
 * @param {string} src @returns {string}
 */
function stripComments(src) {
  return src
    .replace(/\/\*[\s\S]*?\*\//g, (m) => m.replace(/[^\n]/g, ' '))
    .replace(/\/\/[^\n]*/g, '');
}

const violations = [];

for (const file of walk(ROOT)) {
  const rel = file.split(sep).join(posix.sep);
  if (rel.startsWith(ALLOWED_PREFIX)) continue;

  stripComments(readFileSync(file, 'utf8'))
    .split('\n')
    .forEach((line, i) => {
      if (line.includes(NEEDLE)) violations.push(`${rel}:${i + 1}: ${line.trim()}`);
    });
}

if (violations.length > 0) {
  console.error(
    `L4 vi pham: '${NEEDLE}' chi duoc xuat hien trong ${ALLOWED_PREFIX}\n` +
      violations.map((v) => `  ${v}`).join('\n'),
  );
  process.exit(1);
}

console.log(`guard:sql OK — khong co '${NEEDLE}' nao ngoai ${ALLOWED_PREFIX}`);

File: ops/guard_sql.py
"""L4: khong cau SQL nao duoc cham bang memory_fact ngoai memory/repository/.

import-linter so khop DUONG DAN MODULE, khong doc duoc chuoi SQL — hop dong 5 trong
.importlinter chi chan viec IMPORT fact_repo, khong chan ai do viet
'SELECT ... FROM memory_fact' thang trong knowledge/. Day la lop con thieu.

Chay: uv run python ops/guard_sql.py   (bat buoc trong CI)
"""

import ast
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent / "src"
ALLOWED_PREFIX = ("memory", "repository")
NEEDLE = "memory_fact"


def violations() -> list[str]:
    found: list[str] = []

    for path in sorted(ROOT.rglob("*.py")):
        relative = path.relative_to(ROOT)
        if relative.parts[: len(ALLOWED_PREFIX)] == ALLOWED_PREFIX:
            continue

        tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
        # Chi xet CHUOI trong ma nguon. Nhac ten bang trong comment hay docstring la
        # hop le — chay SQL len no thi khong. AST bo comment san, con docstring thi
        # loai o duoi.
        docstrings = {
            id(node.body[0].value)
            for node in ast.walk(tree)
            if isinstance(node, ast.Module | ast.ClassDef | ast.FunctionDef | ast.AsyncFunctionDef)
            and node.body
            and isinstance(node.body[0], ast.Expr)
            and isinstance(node.body[0].value, ast.Constant)
        }

        for node in ast.walk(tree):
            if (
                isinstance(node, ast.Constant)
                and isinstance(node.value, str)
                and NEEDLE in node.value
                and id(node) not in docstrings
            ):
                found.append(f"{relative}:{node.lineno}")

    return found


def main() -> int:
    found = violations()
    if found:
        print(f"L4 vi pham: '{NEEDLE}' chi duoc xuat hien trong memory/repository/")
        for line in found:
            print(f"  {line}")
        return 1
    print(f"guard_sql OK - khong co '{NEEDLE}' nao ngoai memory/repository/")
    return 0


if __name__ == "__main__":
    sys.exit(main())
